### ISSUE TITLE
[FEATURE] Image data archiving and retrieval to keep working set small

### ISSUE THREAD

**Issue by zhill** [anchore/anchore-engine#165]
# Managing Image Analysis Data

This is a summary of the feature. An initial implementation is already done but of course is subject to revision/feedback. This will remain open until the feature is released and will be updated with a reference to the commit that addresses it.

## Part 1: Image Analysis Archiving

Anchore Engine is a data intensive system, and storage consumption grows with the number of images analyzed in the system. Currently, users must manually manage storage consumption by:
1. Over-provisioning storage significantly
2. Increasing capacity over time, resulting in downtime (e.g. stop system, grow the db volume, restart)
3. Manually deleting image analysis to free space as needed

In most cases, option 1 only works for a while, which then requires using 2 or 3. Managing storage provisioned for a postgres DB is somewhat complex and may require significant data copies to new volumes to grow capacity over time.

To help mitigate the storage growth of the db itself, Anchore Engine already provides the 'archive' subsystem that allows usage of external object stores like S3 or Swift to offload the unstructured data storage needs to systems that are more growth tolerant and flexible. This lowers the db overhead but does not fundamentally address the issue of unbounded growth in a busy system.

## Proposed Solution 
In most cases, the model that most users will want to implement is two-tiered storage with a working set and an archived set of images.

### Working Set Images
The set of images as are currently implemented in the system: automatically updated with vulnerability/feed updates and available for queries/policy evaluation. In the current implementation, all images that are analyzed fall into this category.

### Archive Set Images
A new stage of image lifecycle in the system, these images were analyzed and in the working set, but have been transitioned to a new "archived" state. In this state, the image digest and tag mappings are retained in the db but all content/vuln/policy evaluation queries are not longer available and the image is not longer in the /images collection, but rather in the /archives/images collection (this is opposed to a filter since the metadata is different and it helps keep api calls for working set from being impacted at all by the size of the archive set).

### Transitions
Lifecycle transitions are possible from analyzed->archived and archived->analyzed. A user action will transition the image to the new stage and begin an asynchronous process to migrate the image analysis report and the cve/policy evaluation data to the image archive storage location. This archive storage location will be one of two locations:

1. A specific logical bucket-prefix (mapped to backends usually as a key prefix) “analysis_archive” location in the existing archive subsystem (which may be the postgres db if configured as such, though not recommended).
2. A separate image archive location in the configuration, distinct from the existing document archive but using the same configuration parameters. This supports using a different driver (e.g. s3/swift) or a different bucket/container.

#### Archiving an Image Analysis
Images are archived via an API operation: 

POST /archives/images - Body is a list of digest strings to support batch operation, each digest must be the digest of a valid image in the system in the ‘analyzed’ state. The response is a list of status responses, one for each digest provided.

GET /archives/images/<digest> -- Returns the archive metadata (tags, timestamps, etc)

DELETE /archives/images/<digest> -- Deletes the archive data. This means the image may not be restored from archive without additional user action to restore the archive or provide the archive later inline.

#### Restoring an Archived Image
Images may be restored from archive to analyzed state via an API operation:

The image archive content must be available in the image archive first, then call API:

POST /images -- BODY= {“source”: {“archive”: {“digest”: “sha256:abc”}}}

All tag histories will be restored from the archive that are present.

The restore process will re-load the image including its policy evaluation history, but the image will be reloaded into the policy engine as a new image, so vulnerability and policy evaluations after the load will indicate the vuln status of the image as of the current time, not the time when the image was archived.

# Part 2: Automated lifecycle controls

Once an archive process is in place, automated systems to manage archival of images can be added. These mechanisms will provide a sliding-window model on the data that will automatically transition images to the archive based on configuration. It is always possible to restore images from the archive on-demand.

## Defining the Sliding Window

Because the update frequency and deployment rate of images in different applications/repositories can vary widely, the lifecycle rules allow for customization of the sliding window based on both time and tag history. The time window bounds the application of rules and the specific rules bound the amount of tag history maintained in the working set of images.

Rules are defined as follows:
* selector: a json object defining a set of filters on registry, repository, and tag that this rule will apply to. Each entry supports wildcards. e.g. `{"registry": "*", "repository": "library/*", "tag": "latest"}`
* tag_versions_newer: the number of tag->digest mappings with newer timestamps that must be preset for this rule to match an image's tags.
* analysis_age_days: the age of the analysis records (by looking at created_at timestamps). A record at least this number of days old will match
* transition: the operation to perform. if 'archive', works on the working set and transitions to archive, if 'delete' then works on the archive set and deletes the archived record on a match. Note that an 'archive' transition will create an archive of the image and upon success, delete the source analysis to remove it from the working set. This is different from how the direct API works, since that mechanism requires the user to explicitly archive and then delete the working set data. This rule combines those actions.

Example rule in json:
```
{ 
  “selector": {
    "registry": "*",
    "repository": "*",
    "tag": "*"
  },
  “analysis_age_days”: 90, 
  “tag_versions_newer”: 3, 
  “transition”: “archive”
}
```
#### Rule conflicts and application:

Since images may be referenced by multiple tags, an image is only acted upon if all of its tags match one or more rules and result in the same action. The tags do not have to match the same rule, an image may be transitioned by a combination of rules acting on different tag records it has mapped.

**Comment by zhill** [comment]
Initial implementation in beec9010abd628a01f8b332e9649a3560b0cdad7. Some cleanup and refinement to be done, but can see updated swagger for the API in the [external API swagger.yaml](https://github.com/anchore/anchore-engine/blob/master/anchore_engine/services/apiext/swagger/swagger.yaml#L2047)

**Comment by zhill** [comment]
Closing this as it was implemented and release in v0.4.0 of anchore engine.
